File: lib/test_api_integration.dart
import 'dart:io';

import 'package:flutter/material.dart';

import '../models/appointment.dart';
import '../services/api/api_client.dart';
import '../services/api/appointment_service.dart';
import '../services/api/auth_service.dart';
import '../services/api/doctor_service.dart';

/// Simple test to verify API integration
class ApiIntegrationTest {
  static Future<void> runTests() async {
    print('🧪 Starting API Integration Tests...\n');
    
    final apiClient = ApiClient();
    final authService = AuthService();
    final doctorService = DoctorService();
    final appointmentService = AppointmentService();
    
    bool allTestsPassed = true;
    
    // Test 1: Health Check
    print('📋 Test 1: Health Check');
    try {
      final response = await apiClient.get('/health', requireAuth: false);
      print('✅ Health check passed: ${response.body}');
    } catch (e) {
      print('❌ Health check failed: $e');
      allTestsPassed = false;
    }
    
    // Test 2: Authentication Test
    print('\n📋 Test 2: Authentication Test');
    try {
      final result = await authService.testAuth();
      print('✅ Authentication test passed: $result');
    } catch (e) {
      print('❌ Authentication test failed: $e');
      allTestsPassed = false;
    }
    
    // Test 3: Doctor Login
    print('\n📋 Test 3: Doctor Login');
    try {
      final token = await authService.login('doctor@hospital.com', 'password123');
      print('✅ Login successful: Token received');
      print('   Token: ${token.substring(0, 20)}...');
    } catch (e) {
      print('❌ Login failed: $e');
      allTestsPassed = false;
    }
    
    // Test 4: Get Current Doctor
    print('\n📋 Test 4: Get Current Doctor');
    try {
      final doctor = await authService.getCurrentDoctor();
      print('✅ Current doctor retrieved: ${doctor.name}');
      print('   Email: ${doctor.email}');
      print('   Specialization: ${doctor.specialization}');
    } catch (e) {
      print('❌ Get current doctor failed: $e');
      allTestsPassed = false;
    }
    
    // Test 5: Get All Doctors
    print('\n📋 Test 5: Get All Doctors');
    try {
      final doctors = await doctorService.getAllDoctors();
      print('✅ Doctors retrieved: ${doctors.length} doctors found');
      for (final doctor in doctors.take(2)) {
        print('   - ${doctor.name} (${doctor.specialization})');
      }
    } catch (e) {
      print('❌ Get doctors failed: $e');
      allTestsPassed = false;
    }
    
    // Test 6: Get Appointments
    print('\n📋 Test 6: Get Appointments');
    try {
      final appointments = await appointmentService.getAllAppointments();
      print('✅ Appointments retrieved: ${appointments.length} appointments found');
      for (final appointment in appointments.take(2)) {
        print('   - ${appointment.patientDetails.name} with Dr. ${appointment.doctor.name}');
      }
    } catch (e) {
      print('❌ Get appointments failed: $e');
      allTestsPassed = false;
    }
    
    // Test 7: Create Appointment
    print('\n📋 Test 7: Create Appointment');
    try {
      final newAppointment = await appointmentService.createAppointment(
        doctorId: '1',
        date: DateTime.now().add(Duration(days: 2)),
        timeSlot: '02:00 PM',
        patientDetails: PatientDetails(
          name: 'Test Patient',
          age: '35',
          phone: '9876543210',
          symptoms: 'Test symptoms',
        ),
      );
      print('✅ Appointment created successfully');
      print('   ID: ${newAppointment.id}');
      print('   Token: ${newAppointment.token}');
    } catch (e) {
      print('❌ Create appointment failed: $e');
      allTestsPassed = false;
    }
    
    // Test 8: Logout
    print('\n📋 Test 8: Logout');
    try {
      await authService.logout();
      print('✅ Logout successful');
    } catch (e) {
      print('❌ Logout failed: $e');
      allTestsPassed = false;
    }
    
    // Final Result
    print('\n${'=' * 50}');
    if (allTestsPassed) {
      print('🎉 All API integration tests passed!');
      print('✅ Your Flutter app is ready to connect to the backend!');
    } else {
      print('❌ Some tests failed. Please check the errors above.');
    }
    print('=' * 50);
  }
  
  /// Test network connectivity
  static Future<bool> testNetworkConnectivity() async {
    print('🌐 Testing network connectivity...');
    try {
      final result = await InternetAddress.lookup('localhost');
      if (result.isNotEmpty && result[0].rawAddress.isNotEmpty) {
        print('✅ Network connectivity OK');
        return true;
      }
    } catch (e) {
      print('❌ Network connectivity failed: $e');
    }
    return false;
  }
  
  /// Test backend server availability
  static Future<bool> testBackendServer() async {
    print('🔧 Testing backend server availability...');
    try {
      final apiClient = ApiClient();
      final response = await apiClient.get('/health', requireAuth: false);
      if (response.statusCode == 200) {
        print('✅ Backend server is running');
        return true;
      }
    } catch (e) {
      print('❌ Backend server not available: $e');
    }
    return false;
  }
}

/// Widget to run tests in Flutter app
class ApiTestWidget extends StatefulWidget {
  const ApiTestWidget({super.key});

  @override
  _ApiTestWidgetState createState() => _ApiTestWidgetState();
}

class _ApiTestWidgetState extends State<ApiTestWidget> {
  bool _isRunning = false;
  String _result = '';
  
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('API Integration Test'),
      ),
      body: Padding(
        padding: EdgeInsets.all(16.0),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            ElevatedButton(
              onPressed: _isRunning ? null : _runTests,
              child: _isRunning 
                ? Row(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: [
                      SizedBox(
                        width: 20,
                        height: 20,
                        child: CircularProgressIndicator(strokeWidth: 2),
                      ),
                      SizedBox(width: 8),
                      Text('Running Tests...'),
                    ],
                  )
                : Text('Run API Tests'),
            ),
            SizedBox(height: 20),
            Expanded(
              child: Container(
                padding: EdgeInsets.all(16),
                decoration: BoxDecoration(
                  border: Border.all(color: Colors.grey),
                  borderRadius: BorderRadius.circular(8),
                ),
                child: SingleChildScrollView(
                  child: Text(
                    _result.isEmpty ? 'Test results will appear here...' : _result,
                    style: TextStyle(
                      fontFamily: 'monospace',
                      fontSize: 12,
                    ),
                  ),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
  
  Future<void> _runTests() async {
    setState(() {
      _isRunning = true;
      _result = 'Running API integration tests...\n\n';
    });
    
    try {
      await ApiIntegrationTest.runTests();
      setState(() {
        _isRunning = false;
        _result += '\n✅ Tests completed! Check console for detailed results.';
      });
    } catch (e) {
      setState(() {
        _isRunning = false;
        _result += '\n❌ Test execution failed: $e';
      });
    }
  }
}

File: backend/test_api_integration.py
#!/usr/bin/env python3
"""
Comprehensive API integration test
"""
import requests
import json
from datetime import datetime, timedelta

# API Configuration
BASE_URL = "http://localhost:8000"

def test_api_integration():
    """Test all API endpoints"""
    print("🧪 Starting API Integration Tests")
    print("=" * 50)
    
    results = []
    
    # Test 1: Health Check
    print("\n📋 Test 1: Health Check")
    try:
        response = requests.get(f"{BASE_URL}/health", timeout=5)
        if response.status_code == 200:
            print("✅ Health check passed")
            print(f"   Response: {response.json()}")
            results.append(True)
        else:
            print(f"❌ Health check failed: {response.status_code}")
            results.append(False)
    except Exception as e:
        print(f"❌ Health check failed: {e}")
        results.append(False)
    
    # Test 2: Get Doctors
    print("\n📋 Test 2: Get All Doctors")
    try:
        response = requests.get(f"{BASE_URL}/api/doctors/", timeout=5)
        if response.status_code == 200:
            doctors = response.json()
            print(f"✅ Doctors retrieved: {len(doctors)} doctors")
            for doctor in doctors[:2]:
                print(f"   - {doctor['name']} ({doctor['specialization']})")
            results.append(True)
        else:
            print(f"❌ Get doctors failed: {response.status_code}")
            results.append(False)
    except Exception as e:
        print(f"❌ Get doctors failed: {e}")
        results.append(False)
    
    # Test 3: Authentication
    print("\n📋 Test 3: Doctor Login")
    try:
        login_data = {
            "email": "doctor@hospital.com",
            "password": "password123"
        }
        response = requests.post(f"{BASE_URL}/api/auth/login-json", 
                               json=login_data, timeout=5)
        if response.status_code == 200:
            token_data = response.json()
            token = token_data['access_token']
            print("✅ Login successful")
            print(f"   Token: {token[:20]}...")
            results.append(True)
            
            # Test 4: Get Current Doctor with Token
            print("\n📋 Test 4: Get Current Doctor")
            headers = {"Authorization": f"Bearer {token}"}
            response = requests.get(f"{BASE_URL}/api/auth/me", headers=headers, timeout=5)
            if response.status_code == 200:
                doctor = response.json()
                print(f"✅ Current doctor: {doctor['name']}")
                print(f"   Email: {doctor['email']}")
                print(f"   Specialization: {doctor['specialization']}")
                results.append(True)
            else:
                print(f"❌ Get current doctor failed: {response.status_code}")
                results.append(False)
                
            # Test 5: Get Appointments with Token
            print("\n📋 Test 5: Get Appointments")
            response = requests.get(f"{BASE_URL}/api/appointments/", headers=headers, timeout=5)
            if response.status_code == 200:
                appointments = response.json()
                print(f"✅ Appointments retrieved: {len(appointments)} appointments")
                for apt in appointments[:2]:
                    print(f"   - {apt['patient_name']} (Token: {apt['token']})")
                results.append(True)
            else:
                print(f"❌ Get appointments failed: {response.status_code}")
                results.append(False)
                
            # Test 6: Create Appointment
            print("\n📋 Test 6: Create Appointment")
            appointment_data = {
                "doctor_id": "1",
                "date": (datetime.now() + timedelta(days=2)).isoformat(),
                "time_slot": "02:00 PM",
                "patient_details": {
                    "name": "Test Patient",
                    "age": "35",
                    "phone": "9876543210",
                    "symptoms": "Test symptoms"
                }
            }
            response = requests.post(f"{BASE_URL}/api/appointments/", 
                                   json=appointment_data, timeout=5)
            if response.status_code == 200:
                appointment = response.json()
                print(f"✅ Appointment created: {appointment['id']}")
                print(f"   Token: {appointment['token']}")
                results.append(True)
            else:
                print(f"❌ Create appointment failed: {response.status_code}")
                print(f"   Error: {response.text}")
                results.append(False)
                
        else:
            print(f"❌ Login failed: {response.status_code}")
            print(f"   Error: {response.text}")
            results.append(False)
    except Exception as e:
        print(f"❌ Authentication test failed: {e}")
        results.append(False)
    
    # Results Summary
    print("\n" + "=" * 50)
    passed = sum(results)
    total = len(results)
    print(f"📊 Test Results: {passed}/{total} tests passed")
    
    if passed == total:
        print("🎉 All tests passed! API integration is working perfectly!")
    else:
        print("⚠️  Some tests failed. Check the errors above.")
    
    return passed == total

def test_server_connectivity():
    """Test if the server is running"""
    print("🌐 Testing server connectivity...")
    try:
        response = requests.get(f"{BASE_URL}/health", timeout=3)
        if response.status_code == 200:
            print("✅ Server is running and accessible")
            return True
        else:
            print(f"❌ Server responded with status {response.status_code}")
            return False
    except requests.exceptions.ConnectionError:
        print("❌ Cannot connect to server - is it running?")
        print("   Start the server with: python3 -m uvicorn app.main:app --host 0.0.0.0 --port 8000")
        return False
    except Exception as e:
        print(f"❌ Connectivity test failed: {e}")
        return False

if __name__ == "__main__":
    print("🏥 HospitalEasy Backend API Test")
    print("=" * 50)
    
    # First test connectivity
    if test_server_connectivity():
        # Run full integration test
        success = test_api_integration()
        exit(0 if success else 1)
    else:
        print("\n❌ Please start the backend server first:")
        print("   cd /Users/kiranlalk/Desktop/hospitaleasy/backend")
        print("   python3 -m uvicorn app.main:app --host 0.0.0.0 --port 8000")
        exit(1)
